File: 2014 Southeast USA Regionals Division 1/containment.py
import collections
import sys

class Arc:
    def __init__(self, u, rev, cap):
        self.u = u
        self.rev = rev
        self.cap = cap
        self.initial_cap = cap

class FlowNetwork:
    def __init__(self, n):
        self.n = n
        self.network = [[] for _ in range(n)]
        self.dist = [-1] * n
        self.iter = [0] * n
    
    def add_arc(self, u, v, cap_uv, cap_vu=0):
        if u == v:
            return
        
        self.network[u].append(Arc(v, len(self.network[v]), cap_uv))
        self.network[v].append(Arc(u, len(self.network[u]) - 1, cap_vu))
    
    def bfs(self, s, t):
        self.dist = [-1] * self.n
        self.dist[s] = 0
        q = collections.deque([s])
        
        while q:
            v = q.popleft()
            for arc in self.network[v]:
                if arc.cap > 0 and self.dist[arc.u] == -1:
                    self.dist[arc.u] = self.dist[v] + 1
                    q.append(arc.u)
        
        return self.dist[t] != -1
    
    def dfs(self, v, t, flow):
        if v == t:
            return flow
        
        while self.iter[v] < len(self.network[v]):
            arc = self.network[v][self.iter[v]]
            if arc.cap > 0 and self.dist[arc.u] == self.dist[v] + 1:
                f = self.dfs(arc.u, t, min(flow, arc.cap))
                if f > 0:
                    arc.cap -= f
                    self.network[arc.u][arc.rev].cap += f
                    return f
            self.iter[v] += 1
        
        return 0
    
    def max_flow(self, s, t):
        flow = 0
        
        while self.bfs(s, t):
            self.iter = [0] * self.n
            while True:
                f = self.dfs(s, t, float('inf'))
                if f == 0:
                    break
                flow += f
        
        return flow

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    
    n = int(data[0])
    idx = 1
    
    fn = FlowNetwork(1002)
    s, t = 1000, 1001
    
    # Build the 3D grid network
    for i in range(10):
        for j in range(10):
            for k in range(10):
                u = i * 100 + j * 10 + k
                fn.add_arc(u, (i - 1) * 100 + j * 10 + k if i > 0 else t, 1)
                fn.add_arc(u, (i + 1) * 100 + j * 10 + k if i < 9 else t, 1)
                fn.add_arc(u, i * 100 + (j - 1) * 10 + k if j > 0 else t, 1)
                fn.add_arc(u, i * 100 + (j + 1) * 10 + k if j < 9 else t, 1)
                fn.add_arc(u, i * 100 + j * 10 + (k - 1) if k > 0 else t, 1)
                fn.add_arc(u, i * 100 + j * 10 + (k + 1) if k < 9 else t, 1)
    
    # Add source connections
    for _ in range(n):
        x, y, z = int(data[idx]), int(data[idx+1]), int(data[idx+2])
        idx += 3
        fn.add_arc(s, x * 100 + y * 10 + z, 6)
    
    result = fn.max_flow(s, t)
    print(result)

if __name__ == "__main__":
    main()

File: 2014 Southeast USA Regionals Division 1/containment.cs
using System;
using System.Collections.Generic;

public class FlowNetwork {
    public class Arc {
        public int u, rev;
        public int cap, initial_cap;
        
        public Arc(int u, int rev, int cap) {
            this.u = u;
            this.rev = rev;
            this.cap = cap;
            this.initial_cap = cap;
        }
    }
    
    public int n;
    public List<Arc>[] network;
    public int[] dist;
    public int[] iter;
    
    public FlowNetwork(int n) {
        this.n = n;
        network = new List<Arc>[n];
        dist = new int[n];
        iter = new int[n];
        
        for (int i = 0; i < n; i++) {
            network[i] = new List<Arc>();
        }
    }
    
    public void AddArc(int u, int v, int cap_uv, int cap_vu = 0) {
        if (u == v) return;
        
        network[u].Add(new Arc(v, network[v].Count, cap_uv));
        network[v].Add(new Arc(u, network[u].Count - 1, cap_vu));
    }
    
    public bool Bfs(int s, int t) {
        Array.Fill(dist, -1);
        dist[s] = 0;
        
        Queue<int> q = new Queue<int>();
        q.Enqueue(s);
        
        while (q.Count > 0) {
            int v = q.Dequeue();
            foreach (var arc in network[v]) {
                if (arc.cap > 0 && dist[arc.u] == -1) {
                    dist[arc.u] = dist[v] + 1;
                    q.Enqueue(arc.u);
                }
            }
        }
        return dist[t] != -1;
    }
    
    public int Dfs(int v, int t, int flow) {
        if (v == t) return flow;
        
        for (; iter[v] < network[v].Count; iter[v]++) {
            var arc = network[v][iter[v]];
            if (arc.cap > 0 && dist[arc.u] == dist[v] + 1) {
                int f = Dfs(arc.u, t, Math.Min(flow, arc.cap));
                if (f > 0) {
                    arc.cap -= f;
                    network[arc.u][arc.rev].cap += f;
                    return f;
                }
            }
        }
        return 0;
    }
    
    public int MaxFlow(int s, int t) {
        int flow = 0;
        
        while (Bfs(s, t)) {
            Array.Fill(iter, 0);
            int f;
            while ((f = Dfs(s, t, int.MaxValue)) > 0) {
                flow += f;
            }
        }
        return flow;
    }
}

public class Program {
    public static void Main() {
        int n = int.Parse(Console.ReadLine());
        
        FlowNetwork fn = new FlowNetwork(1002);
        int s = 1000, t = 1001;
        
        // Build the 3D grid network
        for (int i = 0; i < 10; i++) {
            for (int j = 0; j < 10; j++) {
                for (int k = 0; k < 10; k++) {
                    int u = i * 100 + j * 10 + k;
                    fn.AddArc(u, i > 0 ? (i - 1) * 100 + j * 10 + k : t, 1);
                    fn.AddArc(u, i < 9 ? (i + 1) * 100 + j * 10 + k : t, 1);
                    fn.AddArc(u, j > 0 ? i * 100 + (j - 1) * 10 + k : t, 1);
                    fn.AddArc(u, j < 9 ? i * 100 + (j + 1) * 10 + k : t, 1);
                    fn.AddArc(u, k > 0 ? i * 100 + j * 10 + (k - 1) : t, 1);
                    fn.AddArc(u, k < 9 ? i * 100 + j * 10 + (k + 1) : t, 1);
                }
            }
        }
        
        // Add source connections
        for (int i = 0; i < n; i++) {
            string[] inputs = Console.ReadLine().Split();
            int x = int.Parse(inputs[0]);
            int y = int.Parse(inputs[1]);
            int z = int.Parse(inputs[2]);
            fn.AddArc(s, x * 100 + y * 10 + z, 6);
        }
        
        int result = fn.MaxFlow(s, t);
        Console.WriteLine(result);
    }
}

// Extension method for Array.Fill for older .NET versions
public static class ArrayExtensions {
    public static void Fill<T>(this T[] array, T value) {
        for (int i = 0; i < array.Length; i++) {
            array[i] = value;
        }
    }
}